### ISSUE TITLE
YouTube API not handling `part` in Playlist Call

### ISSUE THREAD

**Issue by Thiencau20** [Spencer-Easton/Apps-Script-GoogleApis-Libraries#4]
Hello mhawksey

I got this error !

Error: { "error": { "errors": [ { "domain": "youtube.part", "reason": "unknownPart", "message": "snippet?pageToken=CAUQAA", "locationType": "parameter", "location": "part" } ], "code": 400, "message": "snippet?pageToken=CAUQAA" } } (line 83, file "Youtube", project "YouTube")

My code is 👍 

  var goa = cGoa.GoaApp.createGoa('youtube-analytics',
                                  PropertiesService.getUserProperties())
                                  .execute();
   
  YouTube.setTokenService(function(){ return goa.getToken(); } );
  YouTubeAnalytics.setTokenService(function(){ return goa.getToken(); });
  
  //Logger.log('Get Video IDs');
  
  var video = [];
  var playlistId = "PL6kLWSVShvZ38PH1sDoaAfz7k5aMN8uhx";

  var playlistResponde = YouTube.playlistItemsList("snippet",{playlistId:playlistId, maxResults : 5});
  Logger.log(playlistResponde);

How to fix this sir ?

**Comment by Femikuye** [comment]
Hello @Thiencau20 , I am also facing this problem with my script, Please have you found a solution to this problem? If so how did you get it solved? Your Response is Highly appreciated thanks..

**Comment by mhawksey** [comment]
I was able to run your code snippet with my own playlistId and it returned the data without an error. 

Is the issue when you start paging for results? I noticed that error includes a pageToken `snippet?pageToken=CAUQAA` 

Can you share more of you code?

**Comment by mhawksey** [comment]
.. actually I've been able to replicate now. The issue is when there are more results than maxResults. Seeing if I can fix

**Comment by Femikuye** [comment]
Thank you for checking on it, But On my End i have only 2 videos on the channel.. and the maxResults is set to 25.. I guess may be the error is from the library it self.. It seems to me it cant identify the Part.. This is what the error reads.. ``Error: { "error": { "errors": [ { "domain": "youtube.part", "reason": "unknownPart", "message": "snippet?pageToken=CBkQAA", "locationType": "parameter", "location": "part" } ], "code": 400, "message": "snippet?pageToken=CBkQAA" } } (line 83, file "Youtube", project "YouTube")
I don't know what that means but i think you can fix it.

**Comment by mhawksey** [comment]
I think I've fixed - in the script editor if you click Resources > Libraries you should be able to select v2 of both libraries.

**Comment by Femikuye** [comment]
This is my entire code.. I got the error when i run the "retrieveMyUploads()" Functions
`function retrieveMyUploads() {
  var goa = cGoa.GoaApp.createGoa('youtube-analytics',
                                  PropertiesService.getUserProperties())
                                  .execute();
   
  YouTube.setTokenService(function(){ return goa.getToken(); } );
  //YouTubeAnalytics.setTokenService(function(){ return goa.getToken(); });  
 
  var results = YouTube.channelsList('id,snippet,contentDetails,statistics', {mine: true});
  //Logger.log(results[0].contentDetails.relatedPlaylists.uploads);
  //Logger.log(results[0].id);
  var channelId = results[0].id;
  for(var i in results) {
    var item = results[i];
    // Get the playlist ID, which is nested in contentDetails, as described in the
    // Channel resource: https://developers.google.com/youtube/v3/docs/channels
    var playlistId = item.contentDetails.relatedPlaylists.uploads;
    var nextPageToken = '';
    var titleArray = [];
    var idArray = [];
    var reportArray = [];
    
    // This loop retrieves a set of playlist items and checks the nextPageToken in the
    // response to determine whether the list contains additional items. It repeats that process
    // until it has retrieved all of the items in the list.
    while (nextPageToken != null) {
      var playlistResponse = YouTube.playlistItemsList('snippet', {
        playlistId: playlistId,
        maxResults: 25,
        pageToken: nextPageToken
      });

      for (var j = 0; j < playlistResponse.items.length; j++) {
        var playlistItem = playlistResponse.items[j];
        //Logger.log('[%s] Title: %s',
                    //playlistItem.snippet.resourceId.videoId,
                   //playlistItem.snippet.title);
        titleArray[j] = playlistItem.snippet.title;
        idArray[j] = playlistItem.snippet.resourceId.videoId;
        var videoId = playlistItem.snippet.resourceId.videoId;
        reportArray[j] = runYoutubeAnalyticsReport(channelId , videoId)

      }
      //Browser.msgBox("hellon");
      nextPageToken = playlistResponse.nextPageToken;
    }
    create_cheet(idArray , titleArray , reportArray);
  }
}
function runYoutubeAnalyticsReport(channelId , videoId) {
  var goa = cGoa.GoaApp.createGoa('youtube-analytics',
                                  PropertiesService.getUserProperties())
                                  .execute();
   
  //YouTube.setTokenService(function(){ return goa.getToken(); } );
  YouTubeAnalytics.setTokenService(function(){ return goa.getToken(); });  
  var today = new Date();
  var oneMonthAgo = new Date();
  oneMonthAgo.setMonth(today.getMonth() - 1);

  var todayFormatted = Utilities.formatDate(today, 'UTC', 'yyyy-MM-dd');
  var oneMonthAgoFormatted = Utilities.formatDate(oneMonthAgo, 'UTC',
      'yyyy-MM-dd');
  
  // supported dimensions and metrics.
  var analyticsResponse = YouTubeAnalytics.reportsQuery(
    'channel==' + channelId,
    '2017-05-30',
    '2018-11-20',
    'views,subscribersGained,estimatedMinutesWatched,annotationClickThroughRate',
    {
    sort: 'views',
      filters: 'video=='+videoId
    });
  return analyticsResponse
  //create_cheet(analyticsResponse) ;
}
function create_cheet(idArray , titleArray , reportArray){
  // Set the dates for our report
  var today = new Date();
  var oneMonthAgo = new Date();
  oneMonthAgo.setMonth(today.getMonth() - 1);
  var todayFormatted = Utilities.formatDate(today, 'UTC', 'yyyy-MM-dd')
  var oneMonthAgoFormatted = Utilities.formatDate(oneMonthAgo, 'UTC', 'yyyy-MM-dd');
   // Create a new sheet with rows and columns corresponding to our dates
  var sheetName = 'Video ' + oneMonthAgoFormatted + ' - ' + todayFormatted;
  var numRows = reportArray.length;
  var numCols = 6;
  // Get the bound spreadsheet
  var ss = SpreadsheetApp.getActiveSpreadsheet();
 
  // remove any existing sheet for this report
  if (ss.getSheetByName(sheetName)){
    ss.deleteSheet(ss.getSheetByName(sheetName));
  } 
  // insert and set rows/cols
  var sheet = ss.insertSheet(sheetName, {template: ss.getSheetByName('TMP')});
  sheet.insertColumnsAfter(1, numCols-1).insertRowsAfter(1, numRows-1);
 
  // Get the range for the title columns
  // Remember, spreadsheets are 1-indexed, whereas arrays are 0-indexed
  var headersRange = sheet.getRange(1, 1, 1, numCols);
  var headers = ['VideoId','Title','Views','SubscribersGained','AverageWatchTime','ClickThroughRate'];
  var sheet_rows = [];
  for (var j = 0; j < reportArray.length; j++) {
    var myarray = idArray[j]+','+titleArray[j]+','+reportArray[j].rows
    sheet_rows[j] = myarray.split(",");
  }
  // This takes a 2 dimensional array
  headersRange.setValues([headers]);
 
  // Bold and freeze the column names
  headersRange.setFontWeight('bold');
  sheet.setFrozenRows(1);
 
  // Get the data range and set the values
  var dataRange = sheet.getRange(2, 1, numRows, numCols);
  dataRange.setValues(sheet_rows);
 
  // Bold and freeze the dates
  var dateHeaders = sheet.getRange(2, 1, numRows, 1);
  dateHeaders.setFontWeight('bold');
  sheet.setFrozenColumns(1);
 
  // Include the headers in our range. The headers are used
  // to label the axes
  var range = sheet.getRange(1, 1, numRows, numCols);
  var chart = sheet.newChart()
                   .asColumnChart()
                   .setStacked()
                   .addRange(range)
                   .setPosition(4, 2, 10, 10)
                   .build();
  sheet.insertChart(chart);
}`

**Comment by mhawksey** [comment]
Do you still have an issue if you use v2 of the library?

**Comment by Femikuye** [comment]
Thank you.. I see The V2 but my code stay on Running The functions.. It does not stop running!

**Comment by Femikuye** [comment]
When I switched to the v2, my code keep running and not stop!

**Comment by mhawksey** [comment]
Hi - `YouTube.playlistItemsList()` will automatically page all the responses and doesn't return a `.nextPageToken` - as a result I think the while loop never completes. You can test by commenting out your line:

    nextPageToken = playlistResponse.nextPageToken;

**Comment by Femikuye** [comment]
@mhawksey .  It seems your idea is working! when i removed the object propert: " pageToken: nextPageToken" from the YouTube.playlistItemsList() function, the code completes execution though return an error! But I don't understand! Do you mean YouTube.playlistItemsList() will automatically list all the videos in the channel and does not return the nextPageToken?

**Comment by mhawksey** [comment]
@Femikuye correct. The shared functions this library uses for the API calls automatically handles paging https://github.com/Spencer-Easton/Apps-Script-GoogleApis-Libraries/blob/master/Generator-SRC/SharedLibraryFunctions.gs#L61-L73

**Comment by Femikuye** [comment]
Okay Thank you.. Another Issue now is that the YouTube Analytics Libaray you updated is now triggers error.. When i run my App Script code to get YouTube Analytics report i got the error as follows.. Hopw you would be able to detect where the is the error and why does it occur.. Thanks

Error: { "error": { "errors": [ { "domain": "global", "reason": "deleted", "message": "Version 1 of the YouTube Analytics API has been deprecated as of November 1, 2018. Please refer to the migration guide for details about how to update your client to work with version 2 of the API: https://developers.google.com/youtube/analytics/migration." } ], "code": 410, "message": "Version 1 of the YouTube Analytics API has been deprecated as of November 1, 2018. Please refer to the migration guide for details about how to update your client to work with version 2 of the API: https://developers.google.com/youtube/analytics/migration." } } (line 53, file "Code", project "YouTubeAnalytics")

**Comment by mhawksey** [comment]
@Femikuye just added v4 of the library which uses v2 of the YouTube Analytics library. If you open the script editor for your project you should be able to select the newer version via Resources > Library

**Comment by Femikuye** [comment]
Thank You.. I Checked The Resources > Library I can not find the V4 in the YouTube Analytics Library... Do i need to use New Key ?

**Comment by Femikuye** [comment]
Hello, I hope you were able to check on the Library issue?

**Comment by mhawksey** [comment]
... hopefully this time. Should be updated now

**Comment by Femikuye** [comment]
Thank you so much for your effort! Really appreciate it! It is now Updated, but i am getting an eeror while running the YouTube Analytics. This is the error
TypeError: Expected argument of type object, but instead had type string. (line 102, file "Code", project "YouTubeAnalytics")
I think something is still needed to be corrected in the library.. I don't know if am correct! Please check to see the error.. Thank you ;)

**Comment by mhawksey** [comment]
API  changes mean you pass in the query as an object e.g.

 ```
 var analyticsResponse = YouTubeAnalytics.reportsQuery({ids:'channel==' + channelId,
                                                         startDate: oneMonthAgoFormatted,
                                                         endDate: todayFormatted,
                                                         metrics: 'views,likes,dislikes,shares',
                                                         dimensions: 'day',
                                                         sort: '-day'
                                                        });
```

I've updated this in the example

**Comment by Femikuye** [comment]
Yes, it was passed as an object.. I am using the same code i used for the previous library that works..
See my Code Sample Below
 var analyticsResponse = YouTubeAnalytics.reportsQuery(
    'channel==' + channelId,
    '2017-05-30',
    '2018-11-20',
    'views,subscribersGained,estimatedMinutesWatched,annotationClickThroughRate',
    {
    sort: 'views',
      filters: 'video=='+videoId
    });

**Comment by mhawksey** [comment]
that's the old way ... you've got a string followed by an object. It needs to be rewritten as

```
  var analyticsResponse = YouTubeAnalytics.reportsQuery({ids:'channel==' + channelId,
                                                         startDate:'2017-05-30',
                                                         endDate:'2018-11-20',
                                                         metrics:'views,subscribersGained,estimatedMinutesWatched,annotationClickThroughRate',
                                                         sort: 'views',
                                                         filters: 'video=='+videoId
                                                        });
```

**Comment by Femikuye** [comment]
Ooh! I see Thank you.. I just got the point now, thanks very much.. By the way i want to ask if you make the libraries an Open Source ?

**Comment by mhawksey** [comment]
I only generate them from the code @Spencer-Easton has created so up to him :-)

**Comment by Femikuye** [comment]
Okay Thank you for your help. I really Appreciate it.. Thanks a lot ;)

**Comment by SarahGHammam** [comment]
Hi @mhawksey! I am having a similar issue to @Femikuye where my script is stuck at "running function...". This problem occurs when trying to run the following code:

```
var playlistResponse = YouTube.playlistItemsList('snippet', {
        playlistId: playlistId,
        fields: 'items(snippet(publishedAt,resourceId/videoId,title)),nextPageToken'
      });
```
I am using the following libraries: YoutubeAnalytics v3 and Youtube v2 (cannot find any later versions). Can you please help me since I really do not understand what I am doing wrong. 

Many thanks!

----
My entire code looks like this:

```
function retrieveMyUploads() {
  
//https://mashe.hawksey.info/2017/10/getting-youtube-analytics-with-google-apps-script-when-you-get-authentication-loops/ 
  var goa = cGoa.GoaApp.createGoa('youtube-analytics',
                                  PropertiesService.getUserProperties())
                                  .execute();
   
  YouTube.setTokenService(function(){ return goa.getToken(); } );
  YouTubeAnalytics.setTokenService(function(){ return goa.getToken(); });

  
var results = YouTube.channelsList('contentDetails', {mine: true, fields: 'items/contentDetails/relatedPlaylists/uploads'});
  
  // Get the playlist ID, which is nested in contentDetails, as described in the
  // Channel resource: https://developers.google.com/youtube/v3/docs/channels
  var playlistId = results[0].contentDetails.relatedPlaylists.uploads;

      var playlistResponse = YouTube.playlistItemsList('snippet', {
        playlistId: playlistId,
        fields: 'items(snippet(publishedAt,resourceId/videoId,title)),nextPageToken'
      });
      
      Logger.log(playlistResponse);
    }
```

**Comment by SarahGHammam** [comment]
> Hi - `YouTube.playlistItemsList()` will automatically page all the responses and doesn't return a `.nextPageToken` - as a result I think the while loop never completes. You can test by commenting out your line:
> 
> ```
> nextPageToken = playlistResponse.nextPageToken;
> ```

Hi @mhawksey! I am only getting 50 results from the playlistItemsList query. From your comment I assumed I would get all responses without using nextPageToken. How can I get all results?

Thanks in advance!
//Sarah

**Comment by mhawksey** [comment]
Hi @SarahGHammam - it should page automatically. I don't have a playlist with over 50 videos so hard for me to test

**Comment by SarahGHammam** [comment]
Hi again @mhawksey! The code that I am using is the following:

```
var playlistResponse = YouTube.playlistItemsList('snippet, status', {
        playlistId: playlistId,
        maxResults: 50,
        fields: 'items(snippet(publishedAt,resourceId/videoId,thumbnails/default/url,title),status)'
      });
```
I assume now, after some thought, that since I am using the fields parameter (which doesn't include nextPageToken) the paging will not continue. The thing is that when I exclude the field parameter my function gets stuck running and eventually times out. Not really sure how to solve this. Any help is appreciated.

Many thanks!
//Sarah

**Comment by SarahGHammam** [comment]
Hi @mhawksey! Sorry to bother you again. Can I provide more info in order for you to understand my issue? Any help is appreciated.

Thank you!
//Sarah
